src/sagas/tokens.js

```diff
@@ -0,0 +1,255 @@
+import {
+  takeEvery,
+  select,
+  delay,
+  call,
+  fork,
+  take,
+  all,
+  put,
+} from 'redux-saga/effects';
+import { metadataApi } from '@hathor/wallet-lib';
+import { channel } from 'redux-saga';
+import { get } from 'lodash';
+import { TxHistory } from '../models';
+import { specificTypeAndPayload } from './helpers';
+import {
+  types,
+  tokenFetchBalanceSuccess,
+  tokenFetchBalanceFailed,
+  tokenFetchHistorySuccess,
+  tokenFetchHistoryFailed,
+} from '../actions';
+
+export const TOKEN_DOWNLOAD_STATUS = {
+  READY: 'ready',
+  FAILED: 'failed',
+  LOADING: 'loading',
+};
+
+const CONCURRENT_FETCH_REQUESTS = 5;
+const METADATA_MAX_RETRIES = 0;
+
+const mapTokenHistory = (element, token) => {
+  const data = {
+    txId: element.txId,
+    timestamp: element.timestamp,
+    balance: element.balance,
+    // in wallet service this comes as 0/1 and in the full node comes with true/false
+    voided: Boolean(element.voided),
+    tokenUid: token
+  };
+  return new TxHistory(data);
+};
+
+/**
+ * This saga will create a channel to queue TOKEN_FETCH_BALANCE_REQUESTED actions and
+ * consumers that will run in parallel consuming those actions.
+ *
+ * More information about channels can be read on https://redux-saga.js.org/docs/api/#takechannel
+ */
+function* fetchTokenBalanceQueue() {
+  const fetchTokenBalanceChannel = yield call(channel);
+
+  // Fork CONCURRENT_FETCH_REQUESTS threads to download token balances
+  for (let i = 0; i < CONCURRENT_FETCH_REQUESTS; i += 1) {
+    yield fork(fetchTokenBalanceConsumer, fetchTokenBalanceChannel);
+  }
+
+  while (true) {
+    const action = yield take(types.TOKEN_FETCH_BALANCE_REQUESTED);
+    yield put(fetchTokenBalanceChannel, action);
+  }
+}
+
+/**
+ * This saga will consume the fetchTokenBalanceChannel for TOKEN_FETCH_BALANCE_REQUEST actions
+ * and wait until the TOKEN_FETCH_BALANCE_SUCCESS action is dispatched with the specific tokenId
+ */
+function* fetchTokenBalanceConsumer(fetchTokenBalanceChannel) {
+  while (true) {
+    const action = yield take(fetchTokenBalanceChannel);
+
+    yield fork(fetchTokenBalance, action);
+    // Wait until the success action is dispatched before consuming another action
+    yield take(
+      specificTypeAndPayload([
+        types.TOKEN_FETCH_BALANCE_SUCCESS,
+        types.TOKEN_FETCH_BALANCE_FAILED,
+      ], {
+        tokenId: action.tokenId,
+      }),
+    );
+  }
+}
+
+function* fetchTokenBalance(action) {
+  const { tokenId, force } = action;
+
+  try {
+    const wallet = yield select((state) => state.wallet);
+    const tokenBalance = yield select((state) => get(state.tokensBalance, tokenId));
+
+    if (!force && tokenBalance && tokenBalance.oldStatus === 'ready') {
+      // The data is already loaded, we should dispatch success
+      yield put(tokenFetchBalanceSuccess(tokenId, tokenBalance.data));
+      return;
+    }
+
+    const response = yield call(wallet.getBalance.bind(wallet), tokenId);
```

Comment:
Because since the `call` effect will call `getBalance`,  the context will be lost as the function is not pure, so I'm using `bind` to keep the context (so methods from `wallet` can use `this`)